engines/sherlock/events.cpp

```diff
@@ -0,0 +1,299 @@
+
+	_cursorImages = new ImageFile(filename);
+	_cursorId = INVALID_CURSOR;
+}
+
+/**
+ * Set the cursor to show
+ */
+void Events::setCursor(CursorId cursorId) {
+	if (cursorId == _cursorId)
+		return;
+
+	_cursorId = cursorId;
+
+	// Set the cursor data
+	Graphics::Surface &s = (*_cursorImages)[cursorId];
+
+	setCursor(s);
+}
+
+/**
+ * Set the cursor to show from a passed frame
+ */
+void Events::setCursor(const Graphics::Surface &src) {
+	CursorMan.replaceCursor(src.getPixels(), src.w, src.h, 0, 0, 0xff);
+	showCursor();
+}
+
+/**
+ * Show the mouse cursor
+ */
+void Events::showCursor() {
+	CursorMan.showMouse(true);
+}
+
+/**
+ * Hide the mouse cursor
+ */
+void Events::hideCursor() {
+	CursorMan.showMouse(false);
+}
+
+/**
+ * Returns the cursor
+ */
+CursorId Events::getCursor() const {
+	return _cursorId;
+}
+
+/**
+ * Returns true if the mouse cursor is visible
+ */
+bool Events::isCursorVisible() const {
+	return CursorMan.isVisible();
+}
+
+/**
+ * Move the mouse
+ */
+void Events::moveMouse(const Common::Point &pt) {
+	g_system->warpMouse(pt.x, pt.y);
+}
+
+
+/**
+ * Check for any pending events
+ */
+void Events::pollEvents() {
+	checkForNextFrameCounter();
+
+	Common::Event event;
+	while (g_system->getEventManager()->pollEvent(event)) {
+		// Handle keypress
+		switch (event.type) {
+		case Common::EVENT_QUIT:
+		case Common::EVENT_RTL:
+			return;
+
+		case Common::EVENT_KEYDOWN:
+			// Check for debugger
+			if (event.kbd.keycode == Common::KEYCODE_d && (event.kbd.flags & Common::KBD_CTRL)) {
+				// Attach to the debugger
+				_vm->_debugger->attach();
+				_vm->_debugger->onFrame();
+			} else {
+				_pendingKeys.push(event.kbd);
+			}
+			return;
+		case Common::EVENT_KEYUP:
+			return;
+		case Common::EVENT_LBUTTONDOWN:
+			_mouseButtons |= 1;
+			return;
+		case Common::EVENT_RBUTTONDOWN:
+			_mouseButtons |= 2;
+			return;
+		case Common::EVENT_LBUTTONUP:
+			_mouseButtons &= ~1;
+			return;
+		case Common::EVENT_RBUTTONUP:
+			_mouseButtons &= ~2;
+			return;
+		case Common::EVENT_MOUSEMOVE:
+			_mousePos = event.mouse;
+			break;
+		default:
+ 			break;
+		}
+	}
+}
+
+/**
+ * Poll for events and introduce a small delay, to allow the system to
+ * yield to other running programs
+ */
+void Events::pollEventsAndWait() {
+	pollEvents();
+	g_system->delayMillis(10);
+}
+
+/**
+ * Check whether it's time to display the next screen frame
+ */
+bool Events::checkForNextFrameCounter() {
+	// Check for next game frame
+	uint32 milli = g_system->getMillis();
+	if ((milli - _priorFrameTime) >= GAME_FRAME_TIME) {
+		++_frameCounter;
+		_priorFrameTime = milli;
+
+		// Give time to the debugger
+		_vm->_debugger->onFrame();
+
+		// Display the frame
+		_vm->_screen->update();
+
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Get a pending keypress
+ */
+Common::KeyState Events::getKey() {
+	Common::KeyState keyState = _pendingKeys.pop();
+	if ((keyState.flags & Common::KBD_SHIFT) != 0)
+		keyState.ascii = toupper(keyState.ascii);
```

Comment:
Is there any reason why we need to use toupper here? Shouldn't the `ascii` field already contain an upper case character when shift is pressed?